src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs

```diff
@@ -422,7 +422,14 @@ protected sealed override async Task FixAllAsync(Document document, ImmutableArr
                             nodesToRemove.Add(variableDeclarator);
 
                             // Local declaration statement containing the declarator might be a candidate for removal if all its variables get marked for removal.
-                            candidateDeclarationStatementsForRemoval.Add(variableDeclarator.GetAncestor<TLocalDeclarationStatementSyntax>());
+                            // The check for null is necessary for cases where the ancestor tree doesn't contain any TLocalDeclarationStatementSyntax nodes.
+                            // This is the case for variables declared inside a ForStatementSyntax' VariableDeclaration.
+                            // The null would lead to a NPE inside syntaxFacts.GetVariablesOfLocalDeclarationStatement later on when ShouldRemoveStatement passes it along.
+                            var ancestor = variableDeclarator.GetAncestor<TLocalDeclarationStatementSyntax>();
```

Comment:
this doesn't really work.  Consider:

```c#
Action a = () => {
    for (int i = ...
};
```

You can be in a variable declarator that will still be found inside a local declaration statement.

Reply:
Also... my prefernece would be that we just actually support the for-case.   Is there a reason we can't mark and remove the variable inside of a for if it isn't actually used?

Reply:
Also, please add a test for the above case.

Reply:
This should still work, unless i am overlooking somthing:
__________________
### Option 1
The containing lambda is used:
```cs 
    void M()
    {
        Action a = () =>
        {
            for (int [|i|] = 0; ; )
            {
            }
        };
        a();
    }
```
we only populate the candidates in this code. Later on, specifically [here](https://github.com/dotnet/roslyn/blob/dcf4d060f17dae42f3e6ac211a7f6c8436517b09/src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs#L645-L653), we filter out the false positive again because the lambdas body is not unused and therefore not contained in nodesToRemove.
![image](https://user-images.githubusercontent.com/13168856/70949611-821a5580-205e-11ea-9de4-35a285ab6efd.png)
__________________
### Option 2
The containing lambda itself is unused:
```cs
    void M()
    {
        Action a = () =>
        {
            for (int i = 0; ; )
            {
            }
        };
    }
```
In this case we don't actually report a IDE0059 diagnostic on the `i` but rather a CS0219 which would then be fixable by the `CSharpUnusedVariableCodeFixProvider` which has its base class fixed by this PR as well.
![image](https://user-images.githubusercontent.com/13168856/70949628-8d6d8100-205e-11ea-9268-6017e965ed2e.png)

Now this might be a bug in and of itself, which i am not sure about.

For now i added two further tests for both cases above, however i am not sure if i am using the `TestDiagnosticMissingAsync` correctly or if i am supposed to use another helper.

Reply:
This is only working because of external effects we can't guarantee.  It could absolutely be hte case in the future that we report both an unused delegate and an unused for-variable.

We should strive ot make our code actually correct, not pseudo-correct by virtue of how some other external systems work.

Thanks! :)

Reply:
Should this also happen in this PR?
I debugged through the code a bit to actually find why the IDE0059 isn't reported and it looks like the analyzer reports the lambda and then doesn't concern itself with the body any further, presumably because a fix all would actually remove the whole body since it is a non-used value itself.
For a proper fix we would therefore need to step into the lambda and analyze it further. My guess is that the fix should go somwhere around [here](https://github.com/dotnet/roslyn/blob/dcf4d060f17dae42f3e6ac211a7f6c8436517b09/src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs#L183)?

Reply:
> Should this also happen in this PR?

I'm not sure what you mean by 'this'. 

All i'm saying is: the check that walks up arbitrary high in the ancestor clause to see if we're in a local var decl, is incorrect.  We should not do that.  Random cases (like the one i mentioned) would trigger as true, even though they shoudl be false.

The case you're asking is:  is this a variable decl in the foreach statement.  
that's what the code should check for.  it shouldn't infer that by looking to see if it's a local-var-decl or not, as that's not conclusive information.

Reply:
Oh sorry I misunderstood your previous comment. I guess I never questioned the original implementation which I just wrapped in a null check for the fix.